onnx/onnx.proto

```diff
@@ -212,6 +237,44 @@ message TensorProto {
 message SparseTensorProto {
   // The dimensions in the tensor.
   repeated int64 dims = 1;
+  // This field MUST be present this version of the IR.
   optional TensorProto indices = 2;
+  // This field MUST be present this version of the IR.  
   optional TensorProto values = 3;
 }
+
+// Define the types.
+message TypeProto {
+  message TensorTypeProto {
+  // This field MUST NOT have the value of UNDEFINED
+  // This field MUST be present for this version of the IR.
+    optional TensorProto.DataType elem_type = 1;
+  }
+
+  message SparseTensorTypeProto {
+  // This field MUST NOT have the value of UNDEFINED
+  // This field MUST be present for this version of the IR.
+    optional TensorProto.DataType elem_type = 1;
+  }
+
+  oneof value {
```

Comment:
Why do we need a separation (especially of "oneof" kind) for sparse vs dense tensors? Element type is present in either Tensor or SparseTensor.values. Maybe put TensorProto.DataType as member of ArgInfoProto directly?

Or do you imagine some other fields appearing in TypeProto struct?

Reply:
See #55 for other TypeProto variants, which are needed to properly schematize attribute values (which can be scalars or sequences, not just tensors)

.

Reply:
Yes, TypeProto is an extension point, and we would like to have the flexibility to extend it t introduce more data types in the future.